### PR TITLE
ARROW-3988: [C++] Do not build unit tests by default, fix building Gandiva unit tests when ARROW_BUILD_TESTS=OFF

### PR DESCRIPTION
**Pull request by wesm**
I found while working on this that disabling `ARROW_GANDIVA_BUILD_TESTS` would break the build -- I think this was caused by some other changes I made. We should remove that option and instead use the new modular build targets and invoke unit tests using labels. So we would write

```
ninja gandiva  # this will build all libraries and unit tests when ARROW_BUILD_TESTS=ON
ctest -L gandiva
```

### CODE REVIEW

**Comment by wesm** [comment]
I opened https://issues.apache.org/jira/browse/ARROW-3994. That can be done in a follow up patch

File: cpp/README.md [hunk 1]
```diff
@@ -261,7 +265,7 @@ The optional `gandiva` libraries and tests can be built by passing
 `-DARROW_GANDIVA=on`.
 
 ```shell
-cmake .. -DARROW_GANDIVA=on
+cmake .. -DARROW_GANDIVA=ON -DARROW_BUILD_TESTS=ON
```
Comment (kou) [inline cpp/README.md]: Should this be `-DARROW_GANDIVA_BUILD_TESTS=ON`?

**Review by kou — APPROVED** [review]
+1

File: cpp/README.md (hunk 1 above)
Comment (wesm) [reply cpp/README.md]: Indeed. Will fix

File: docs/source/cpp/api/builder.rst [hunk 2]
```diff
@@ -31,9 +31,6 @@ Concrete builder subclasses
 .. doxygenclass:: arrow::BooleanBuilder
    :members:
 
-.. doxygenclass:: arrow::PrimitiveBuilder
```
Comment (xhochy) [inline docs/source/cpp/api/builder.rst]: Is this expected to be here?

Comment (wesm) [reply docs/source/cpp/api/builder.rst]: This was causing a Breath warning and failing the CI build. `PrimitiveBuilder` is an implementation detail anyway (probably it should be in a `details` or `internal` namespace)

**Pull request closed without merging**